### ISSUE TITLE
InvalidCastException calling AddIceCandidate

### ISSUE THREAD

**Issue by AnthonyM** [webrtc-uwp/webrtc-uwp-sdk#132]
I get the information for an ICE Candidate, 

```
var init = new RTCIceCandidateInit()
{
  Candidate = response.Sdp,
  SdpMid = response.SdpMid,
  SdpMLineIndex = (ushort) response.SdpIndex
};
var candidate = new RTCIceCandidate(init);
await _connection.AddIceCandidate(icandidate);

The last line throws an InvalidCastException.

I'm having a lot of trouble debugging this because I'm using 1.71.0.1-Alpha because I need the Setup method which allows disabling A/V capture (as described here https://github.com/webrtc-uwp/webrtc-uwp.github.io/blob/James/20190225-m71-docs/how-to-device-capture-removal.md).

I can't find the code to checkout to build a debug library and start digging into this. Could someone point me to the branch/commit to check out?

**Comment by jamescadd** [comment]
Hi Anthony, you've hit a documentation issue there where I haven't had time to add proper instructions.  If you upgrade to the latest NuGet package (.0.4) you can find the PDBs here: https://1drv.ms/f/s!AtgCZyBqq2SY3MBCBUFoB0vW3VEMQg

You would then need to find some guidance on using loose PDBs to step through the code (I haven't done that myself yet so not much help).  I have read that sometimes it's important for the file location to match the original.  If you run into this, I built these from `C:\repos\webrtc-m71\`.

If you want to do a full build of the source and then link to it from your project just follow the releases/m71 branch instructions for this repo.  Let me know how it goes and I can add your findings to the docs.

**Comment by AnthonyM** [comment]
The m71 branch doesn't have a Setup method which matches the signature of the one I used to disable Audio and Video. 

I thought maybe I could use the WebRtcLibConfiguration to set it up, but now I'm getting an application crash in the native code somewhere. Not sure why yet.

**Comment by jamescadd** [comment]
That signature was changed - here is the updated documentation:
https://github.com/webrtc-uwp/webrtc-uwp.github.io/blob/releases/m71/how-to-device-capture-removal.md

@msplants did you run into an issue like this when disabling audio/video?  I wonder if this needs one of your fixes.

**Comment by AnthonyM** [comment]
Ah okay, thanks. I've got the debug build working now I think. I'll see what I can find out about the error now.

**Comment by AnthonyM** [comment]
It gets back a false result from native_->AddIceCandidate in wrapper::impl::org::webRtc::RTCPeerConnection::addIceCandidate

I can't really go any further because the code becomes completely impossible for me to follow.

**Comment by jamescadd** [comment]
@robin-raymond @morosev could you assist?  @AnthonyM if you share a more full callstack or a repro project it may help.  Also VS output logs for a repro (enable mixed-mode debugging).

**Comment by robin-raymond** [comment]
Yes, I really would need a call stack of some kind of a project that I can reproduce the issue. But I can help if I get that.

**Comment by AnthonyM** [comment]
I don't know if I can get a useful stack trace out of this. The only exception seems to occur in the c# layer.

I did go down through the native layer

In wrapper::impl::org::webRtc::RTCPeerConnection::addIceCandidate
```
  if (!result) return zsLib::Promise::createRejected(UseWebrtcLib::delegateQueue());
```
Result is false, so it uses that branch. I can't really figure out the code beyond that.

Looking at the candidate, I'm seeing that the address is a local address that my client can't connect to (different network). Maybe that's it? That would make sense as a reason for an error, but it's pretty hard to figure out that's the problem.

**Comment by AnthonyM** [comment]
Well that theory wasn't right, I tried making sure only the valid network was available and it still doesn't work.

**Comment by robin-raymond** [comment]
If that is returning false then it's almost certainly that webrtc code is not liking the candidate structure value being passed in and it can be one of the `return false` reasons below:
`
bool PeerConnection::AddIceCandidate(
    const IceCandidateInterface* ice_candidate) {
  TRACE_EVENT0("webrtc", "PeerConnection::AddIceCandidate");
  if (IsClosed()) {
    RTC_LOG(LS_ERROR) << "AddIceCandidate: PeerConnection is closed.";
    NoteAddIceCandidateResult(kAddIceCandidateFailClosed);
    return false;
  }

  if (!remote_description()) {
    RTC_LOG(LS_ERROR) << "AddIceCandidate: ICE candidates can't be added "
                         "without any remote session description.";
    NoteAddIceCandidateResult(kAddIceCandidateFailNoRemoteDescription);
    return false;
  }

  if (!ice_candidate) {
    RTC_LOG(LS_ERROR) << "AddIceCandidate: Candidate is null.";
    NoteAddIceCandidateResult(kAddIceCandidateFailNullCandidate);
    return false;
  }

  bool valid = false;
  bool ready = ReadyToUseRemoteCandidate(ice_candidate, nullptr, &valid);
  if (!valid) {
    NoteAddIceCandidateResult(kAddIceCandidateFailNotValid);
    return false;
  }

  // Add this candidate to the remote session description.
  if (!mutable_remote_description()->AddCandidate(ice_candidate)) {
    RTC_LOG(LS_ERROR) << "AddIceCandidate: Candidate cannot be used.";
    NoteAddIceCandidateResult(kAddIceCandidateFailInAddition);
    return false;
  }

  if (ready) {
    bool result = UseCandidate(ice_candidate);
    if (result) {
      NoteUsageEvent(UsageEvent::REMOTE_CANDIDATE_ADDED);
      NoteAddIceCandidateResult(kAddIceCandidateSuccess);
    } else {
      NoteAddIceCandidateResult(kAddIceCandidateFailNotUsable);
    }
    return result;
  } else {
    RTC_LOG(LS_INFO) << "AddIceCandidate: Not ready to use candidate.";
    NoteAddIceCandidateResult(kAddIceCandidateFailNotReady);
    return true;
  }
}
`

**Comment by robin-raymond** [comment]
````
zsLib::Promise::createRejected(UseWebrtcLib::delegateQueue());
````
^ For this code, that just tells the C# layer that the async task operation has failed and will cause the async routine to throw an exception. It's kind of a generic failure. The webrtc layer doesn't report anything to the upper layer except "false" so it's kind of passed along the stack (which in this case is an async operation).

**Comment by robin-raymond** [comment]
You can set break points inside the peerconnection.cc file if you enable mixed mode debugging. That will likely help you track down what is wrong with the value being sent into that method.
